app/storage/metadata_parser.py

```diff
@@ -1,103 +1,90 @@
+        return items
 
-def validate_metadata(claims, required_metadata):
-    _validate_metadata_values_are_valid(claims, required_metadata)
+    def __repr__(self):
+        return '<MetadataSchema({!r})>'.format(self.__dict__)
 
 
-def _validate_metadata_values_are_valid(claims, required_metadata):
+def validate_marshmallows(claims, questionnaire_specific_metadata):
     """
-    Validate metadata from the JWT claims that are required by the schema/runner.
-    Ensures that the values adhere to the expected format.
-    :param claims:
-    :param required_metadata:
+    Validate a set of claims against the mandatory runner claims along with
+    any questionnaire specific claims.
     """
-    try:
-        for metadata_field in required_metadata:
-            name = metadata_field['name']
-            claim = claims.get(name)
-            if name not in claims:
-                raise InvalidTokenException(
-                    'Missing required key {} from claims'.format(name)
+    dynamic_fields = {}
+
+    for metadata_field in questionnaire_specific_metadata:
+        field_arguments = {}
+        validators = []
+        if metadata_field.get('optional'):
+            field_arguments['required'] = False
+        if metadata_field.get('length'):
+            length_requirements = metadata_field['length']
+            validators.append(
+                validate.Length(
+                    min=length_requirements.get('min'),
+                    max=length_requirements.get('max'),
+                    equal=length_requirements.get('equal'),
                 )
+            )
 
-            logger.debug('parsing metadata', key=name, value=claim)
-            VALIDATORS[metadata_field['validator']](claim)
-
-    except (RuntimeError, ValueError, TypeError) as error:
-        logger.error('Unable to parse metadata', key=name, value=claim, exc_info=error)
-        raise InvalidTokenException(
-            'incorrect data in token for {}'.format(name)
-        ) from error
-    except KeyError as key_error:
-        error_msg = 'Invalid validator for schema metadata - {}'.format(
-            key_error.args[0]
+        dynamic_fields[metadata_field['name']] = VALIDATORS[metadata_field['type']](
+            validate=validators, **field_arguments
         )
-        logger.error(error_msg, exc_info=key_error)
-        raise KeyError(error_msg) from key_error
+
+    questionnaire_metadata_schema = type(
+        'QuestionnaireMetadataSchema', (MetadataSchema,), dynamic_fields
+    )(unknown=INCLUDE)
+
+    # The load method performs validation.
+    return questionnaire_metadata_schema.load(claims)
+
+
+def validate_runner_claims(claims: Dict):
+    """ Validate claims required for runner to function"""
+    return validate_marshmallows(claims, [])
+
+
+def validate_questionnaire_claims(claims: Dict, questionnaire_metadata: List):
+    """ Validate any survey specific claims required for a questionnaire"""
```

Comment:
Doesn't this validate all claims, i.e. including the runner claims, as `QuestionnaireMetadataSchema` is based on `MetadataSchema`?